### PR TITLE
#9630: assert out concat when concatenating along padded dimensions

### PR DESCRIPTION
**Pull request by sjameelTT**
### Ticket
#9630 

### Problem description
Padding is not accounted for in concat. Assert out as it's not possible to support in the current kernels (trivially). Allow padding on the last tensor.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by ntarafdar — APPROVED** [review]

**Comment by sjameelTT** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/10532167350

**Review by yan-zaretskiy — APPROVED** [review]

File: ttnn/cpp/ttnn/tensor/types.hpp [hunk 1]
```diff
@@ -681,6 +684,20 @@ struct Shape {
         return Shape{tt::tt_metal::Shape{this->value, tt::tt_metal::Padding{this->value.rank()}}};
     }
 
+    bool has_tile_padding() const {
+        auto rank = this->rank();
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/tensor/types.hpp:688]: you can simplify a little bit this code to something like this:
bool has_tile_padding() const {
    return std::any_of(
        std::begin(*this), 
        std::begin(*this) + this->rank(), 
        [this](int index) { return this->has_tile_padding(index); }
    );
}

File: ttnn/cpp/ttnn/tensor/types.hpp [hunk 2]
```diff
@@ -407,206 +396,220 @@ struct BorrowedStorage {
...
-                buffers.insert({device->id(), buffer});
-            }
-            std::lock_guard<std::mutex> lock(shape_mtx);
-            shapes.insert({device->id(), shape});
-        }
+    // Helper Functions - Getters and setters to get/modify storage attributes. These are needed to
+    // preinitialize empty tensor handles and use/populate them in the worker threads.
 
-        inline DeviceBuffer get_buffer_for_device(Device* device) const {
+    inline void insert_buffer_and_shape_for_device(Device *device, const DeviceBuffer buffer, const Shape shape) {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/tensor/types.hpp:554]: const DeviceBuffer& buffer.
const Shape& shape

File: ttnn/cpp/ttnn/tensor/types.hpp [hunk 3]
```diff
@@ -407,206 +396,220 @@ struct BorrowedStorage {
...
-                .buffer_type = this->buffers.at(first_device_id)->buffer_type(),
-                .shard_spec = shard_spec};
+    bool operator==(const MultiDeviceStorage &other) {
+        return this->ordered_device_ids == other.ordered_device_ids and this->strategy == other.strategy and
+               this->buffers == other.buffers and this->shapes == other.shapes;
+    }
 
+    inline const MemoryConfig memory_config() const {
+        std::lock_guard<std::mutex> lock(buffer_mtx);
+        auto first_device_id = this->ordered_device_ids.at(0);
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/tensor/types.hpp:534]: overall I prefer do manual check
if (this->ordered_device_ids.size() == 0) {
TT_FATAL("no such device...")
}
auto first_device_id = this->ordered_device_ids[0];
In this case we throw much better error message.

File: ttnn/cpp/ttnn/tensor/types.hpp [hunk 4]
```diff
@@ -407,206 +396,220 @@ struct BorrowedStorage {
...
-        }
-
-        OwnedBuffer get_buffer(int buffer_index) const {
-            std::lock_guard<std::mutex> lock(mtx);
-            TT_ASSERT(buffer_index < buffers.size(), "Buffer not found for buffer_index " + std::to_string(buffer_index));
-            return buffers[buffer_index];;
-        }
+    // Helper Functions - Getters and setters to get/modify storage attributes. These are needed to
+    // preinitialize empty tensor handles and use/populate them in the worker threads.
+    void insert_buffer_and_shape_for_device(int buffer_index, const OwnedBuffer buffer, const Shape shape) {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/tensor/types.hpp:446]: const OwnerBuffer& buffer, const Shape& shape
We don't want to increae decrease atomic counters every call.

File: ttnn/cpp/ttnn/tensor/types.hpp [hunk 5]
```diff
@@ -407,206 +396,220 @@ struct BorrowedStorage {
...
+        buffers[buffer_index] = buffer;
+        shapes[buffer_index] = shape;
+    }
 
-        OwnedBuffer& get_buffer(int buffer_index) {
-            std::lock_guard<std::mutex> lock(mtx);
-            TT_ASSERT(buffer_index < buffers.size(), "Buffer not found for buffer_index " + std::to_string(buffer_index));
-            return buffers[buffer_index];;
-        }
+    OwnedBuffer get_buffer(int buffer_index) const {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/tensor/types.hpp:452]: Did you lost reference? was it intentional?

File: ttnn/cpp/ttnn/tensor/types.hpp [hunk 6]
```diff
@@ -407,206 +396,220 @@ struct BorrowedStorage {
...
+            "Mismatch between device derived from buffer and device derived from MultiDeviceStorage.");
+        return buffers.at(device->id());
+    }
 
-        inline Shape get_tensor_shape_for_device(Device* device) const {
-            std::lock_guard<std::mutex> lock(shape_mtx);
-            TT_ASSERT(shapes.find(device->id()) != shapes.end(), "Shape not found for device " + std::to_string(device->id()));
-            return shapes.at(device->id());
-        }
+    inline DeviceBuffer get_buffer_for_device_id(uint32_t device_id) const {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/tensor/types.hpp:586]: Same I feel like we cshould return const reference here.

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.cpp [hunk 7]
```diff
@@ -39,6 +40,15 @@ void ConcatDeviceOperation::validate(const std::vector<Tensor> &input_tensors) c
         tt::tt_metal::Shape curr_shape = in_ref.get_legacy_shape();
         TT_FATAL(curr_shape.rank() == shape_first.rank(), "Input tensor ranks must be equal");
         curr_shape[this->dim] = 0;
+        if (i < input_tensors.size() - 1) {
+            // last tensor can support without any kernel changes
+            TT_FATAL(
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.cpp:45]: Where is the test for this case?

**Review by Denys88 — APPROVED** [review]

**Pull request merged**